### ISSUE TITLE
Comprehensive Benchmarking Suite

### ISSUE THREAD

**Issue by SkanderMulder** [SkanderMulder/graniteR#14]
# Comprehensive Benchmarking Suite

## Overview
Establish a rigorous benchmarking framework to validate graniteR's performance claims and guide optimization efforts.

## Motivation
- Provide evidence for "9/10 efficiency" claims
- Compare against Python alternatives (sentence-transformers, transformers)
- Identify performance bottlenecks
- Track performance regressions
- Guide hardware recommendations

### 1. Throughput Benchmarks
Measure documents/second for various operations:

```r
# Embedding throughput
benchmark_embed(
  n_samples = c(100, 1000, 10000),
  devices = c("cpu", "cuda"),
  models = c("ibm-granite/granite-embedding-r2", 
             "sentence-transformers/all-MiniLM-L6-v2")
)

# Classification throughput
benchmark_classify(
  n_samples = c(100, 1000, 10000),
  num_labels = c(2, 6, 20),
  methods = c("frozen", "fine-tuned", "moe")
)
```

### 2. Accuracy Benchmarks
Compare model quality on standard datasets:

- **Sentiment**: IMDb, SST-2
- **Topic**: AG News, DBpedia
- **Intent**: CLINC150, Banking77
- **Safety**: ToxicComments, Jigsaw

Metrics:
- Accuracy, F1-macro, F1-weighted
- Confusion matrices
- Per-class performance

### 3. Memory Benchmarks
Profile memory usage:

```r
benchmark_memory(
  operations = c("embed", "train", "predict"),
  batch_sizes = c(8, 16, 32, 64),
  models = c("granite-r2", "bert-base")
)
```

### 4. Training Efficiency
Compare training approaches:

- Frozen backbone vs fine-tuning
- Different learning rates
- Epochs to convergence
- Transfer learning effectiveness (small data)

### 5. AutoML Performance
Evaluate `auto_classify()`:

- Time to best model
- Final accuracy vs baselines
- Ensemble improvement
- Meta-learning effectiveness

### Directory Structure
```
benchmarks/
├── data/
│   ├── download_datasets.R      # Fetch standard datasets
│   └── datasets.rds             # Cached datasets
├── scripts/
│   ├── bench_throughput.R
│   ├── bench_accuracy.R
│   ├── bench_memory.R
│   └── bench_automl.R
├── results/
│   ├── throughput_results.csv
│   ├── accuracy_results.csv
│   └── plots/
├── reports/
│   └── benchmark_report.Rmd    # Automated report generation
└── README.md
```

### Benchmark Functions
```r
# In R/benchmark.R (or separate package graniteR.bench)

#' Run standard benchmark suite
#' @export
run_benchmarks <- function(
  suite = c("quick", "standard", "comprehensive"),
  devices = c("cpu", "cuda"),
  output_dir = "benchmarks/results"
) {
  # ...
}

#' Compare against Python baselines
#' @export
compare_python <- function(
  dataset,
  python_libs = c("sentence-transformers", "transformers")
) {
  # Uses reticulate to run equivalent Python code
  # Returns comparison tibble
}
```

### Visualization
```r
# Performance comparison plots
plot_benchmark_results(results, metric = "throughput")
plot_benchmark_results(results, metric = "accuracy")
plot_benchmark_results(results, metric = "memory")

# Model comparison heatmap
plot_model_comparison(
  models = c("granite-r2", "bert-base", "distilbert"),
  datasets = c("imdb", "ag_news", "sst2")
)
```

### Performance Targets (CPU, Granite R2)
- Embed 1K docs: <30 seconds
- Embed 10K docs: <5 minutes
- Train classifier (1K samples, 3 epochs): <2 minutes
- Predict 1K docs: <20 seconds

### Accuracy Targets (vs Python baselines)
- Within 1% accuracy on standard datasets
- Ensemble improves by 1-3%
- AutoML achieves 90%+ of manual tuning performance

### Memory Targets
- Peak memory <2GB for 10K document embedding
- Training memory scales linearly with batch size
- No memory leaks in long-running processes

### GitHub Actions Workflow
```yaml
# .github/workflows/benchmark.yaml
name: Continuous Benchmarking
on:
  push:
    branches: [main]
  schedule:
    - cron: '0 0 * * 0'  # Weekly

jobs:
  benchmark:
    runs-on: ubuntu-latest
    steps:
      - name: Run benchmarks
        run: Rscript benchmarks/scripts/run_all.R
      - name: Upload results
        uses: actions/upload-artifact@v2
      - name: Comment on PR
        # Post results as PR comment
```

### Regression Detection
- Automatic alerts if performance drops >5%
- Historical tracking of metrics
- Integration with GitHub Pages for result visualization

## Deliverables

1. **Benchmark Package** (or vignette)
   - Documented benchmark functions
   - Standard datasets
   - Automated reporting

2. **Public Results**
   - Hosted on GitHub Pages
   - Updated weekly/monthly
   - Interactive visualizations

3. **Comparison Matrix**
   - graniteR vs sentence-transformers
   - graniteR vs transformers
   - Different hardware (CPU, GPU, M1/M2)

4. **Optimization Guide**
   - Hardware recommendations
   - Batch size tuning
   - Model selection guidance

## Related Issues
- Batch tokenization optimization (#[TBD])
- Performance profiling (#[TBD])
- CRAN preparation (#[TBD])

## Priority
High - Critical for validating performance claims and CRAN release

## Timeline
- Phase 1 (Basic benchmarks): 2 weeks
- Phase 2 (Automation + CI): 2 weeks
- Phase 3 (Public dashboard): 1 week
